### PR TITLE
Add vaccine available parameter to strategies

### PR DESCRIPTION
**Pull request by giovannic**
The interesting code is in src/strategy.js.

I don't remember all the caveats you mentioned on the call about reducing the coverage :/. Let me know if I've fallen into any of them.

Let me know if I can demistify any JS too.

But I'll do a little pre-release early for Andy ;)

Thanks!
Giovanni

### CODE REVIEW

File: src/strategy.js [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+  ) / wholePopulation;
+
+  //create copy of the matrix
+  let scaled = matrix.map(row => row.slice());
+
+  if (currentCoverage > targetCoverage) {
+    const ratio = targetCoverage / currentCoverage;
+    for (let i = 0; i < scaled.length; i++) {
+      for (let j = 0; j < scaled[i].length; j++) {
+        scaled[i][j] *= ratio;
```
Comment (pwinskill) [inline src/strategy.js]: I think this line might invalidate one of the caveats I mentioned.
The age-groups (columns) in the row of the matrix we need to adjust (the first one where currentCoverage>targetCoverage) can't all be rescaled. It should only be those age groups that were increased in that prioritisation step.

I've mocked up an example in R (sorry - not confident enough to do it in JS):
```

matrix <- matrix(c(0, 0, 0.8, 0, 0.8, 0.8), nrow = 2, byrow = TRUE)
population <- c(100, 100 , 100)
vaccinesAvailable <- 100

wholePopulation = sum(population)
targetCoverage = vaccinesAvailable / wholePopulation

# Estimate coverage for each step
currentCoverage = c()
for(i in 1:nrow(matrix)){
  currentCoverage[i] <- sum(matrix[i,] * population) / wholePopulation
}
# Identify where we exceed the target
index <- which(currentCoverage > targetCoverage)[1]

# Additonal number of people we need to vaccianted
target = vaccinesAvailable - (currentCoverage[index - 1] * wholePopulation)
# Actual additional people vaccinated in this step
actual = (currentCoverage[index] * wholePopulation) - (currentCoverage[index - 1] * wholePopulation)
# Ratio of target to actual
ratio = target / actual

# Identify the age group(s) that have increased in this step
stepped <- matrix[index,] > matrix[index - 1, ]

# Create the adjsuted matrix
adjustedMatrix <- matrix
adjustedMatrix[index,stepped] <- adjustedMatrix[index,stepped] * ratio

# Check the adusted matrix - perhaps these could be tests?
  # All rows except last are the same
all(adjustedMatrix[-nrow(adjustedMatrix), ] == matrix[-nrow(matrix), ])
  # Coverage of final row == targetCoverage
(sum(adjustedMatrix[nrow(adjustedMatrix),] * population) / wholePopulation) == targetCoverage
  # All columns a monotonically increasing
all(apply(adjustedMatrix, 2, function(x){
  all(diff(x) >= 0)
}))
```

Hopefully that makes sense?! It probably needs some checking on a few test cases once implemented - I'm not sure its 100% robust if there are mutiple steps in a single age-group

**Review by pwinskill — COMMENTED** [review]
Hi Giovanni,
Thanks for this - I think we might need a slight tweak which I've detailed in the specific comment

File: src/strategy.js (hunk 1 above)
Comment (pwinskill) [reply src/strategy.js]: (apologies if I've mis-read the JS and that feature is in there)

Comment (pwinskill) [reply src/strategy.js]: Having thought about this overnight, I can imagine the above failing when more than 1 age groups is increased in the same step. I think this following adjustment solves this:
```

# matrix <- matrix(c(0, 0, 0.8, 0, 0.8, 0.8), nrow = 2, byrow = TRUE)
matrix <- matrix(c(0, 0, 0.8, 0, 0.8, 0.81), nrow = 2, byrow = TRUE)

population <- c(100, 100 , 100)
vaccinesAvailable <- 100

wholePopulation = sum(population)
targetCoverage = vaccinesAvailable / wholePopulation

# Estimate coverage for each step
currentCoverage = c()
for(i in 1:nrow(matrix)){
  currentCoverage[i] <- sum(matrix[i,] * population) / wholePopulation
}
# Identify where we exceed the target
index <- which(currentCoverage > targetCoverage)[1]

# Change in coverage
delta_cov <- matrix[index,] - matrix[index - 1,]
delta_cov_nor <- delta_cov / sum(delta_cov)

# Additional number of people we need to vaccinate
target = delta_cov_nor * (vaccinesAvailable - (currentCoverage[index - 1] * wholePopulation))

# Create the adjsuted matrix
adjustedMatrix <- matrix
adjustedMatrix[index,] <- adjustedMatrix[index - 1,] + (target / population)

# Check the adusted matrix - perhaps these could be tests?
# All rows except last are the same
all(adjustedMatrix[-nrow(adjustedMatrix), ] == matrix[-nrow(matrix), ])
# Coverage of final row == targetCoverage
(sum(adjustedMatrix[nrow(adjustedMatrix),] * population) / wholePopulation) == targetCoverage
# All columns a monotonically increasing
all(apply(adjustedMatrix, 2, function(x){
  all(diff(x) >= 0)
}))
```

**Comment by giovannic** [comment]
Thanks for the feedback. The latest changes include:

 * The scaling code to only edits the last step
 * There's a test that the rest of the matrix is unchanged
 * We work in number of vaccines instead of coverage (this is to save a bunch of divisions and probably some precision errors in JavaScript)

**Review by pwinskill — APPROVED** [review]
Thanks Giovanni. Looks goods - we will have to beta-test the webtool thoroughly to check it is behaving, but hopefully this will address the issue!

**Pull request merged**